File: Easy/0706-design-hashmap.cpp
const int M = 769;

class MyHashMap {
public:
    /** Initialize your data structure here. */
    MyHashMap() {

    }
    
    /** value will always be non-negative. */
    void put(int key, int value) {
        int code = hashCode(key);
        auto it = find_if(lst[code].begin(), lst[code].end(), [key](const pair<int, int>& p) {
            return p.first == key;
        });
        if (it != lst[code].end()) {
            it->second = value;
        } else {
            lst[code].emplace_front(key, value);
        }
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    int get(int key) {
        int code = hashCode(key);
        auto it = find_if(lst[code].begin(), lst[code].end(), [key](const pair<int, int>& p) {
            return p.first == key;
        });
        if (it != lst[code].end()) {
            return it->second;
        }
        return -1;
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    void remove(int key) {
        lst[hashCode(key)].remove_if([key](const pair<int, int>& p) {
            return p.first == key;
        });
    }

private:
    forward_list<pair<int, int>> lst[M];
    
    int hashCode(int x) {
        return x % M;
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */

File: Algorithms/HashTable/0706-design-hashmap.py
class MyHashMap:
    BUCKET_SIZE = 3331

    def __init__(self):
        self.buckets = [[] for _ in range(MyHashMap.BUCKET_SIZE)]


    def getBucket(self, key: int) -> list:
        return self.buckets[key % MyHashMap.BUCKET_SIZE]


    def put(self, key: int, value: int) -> None:
        bucket = self.getBucket(key)
        for element in bucket:
            if element[0] == key:
                element[1] = value
                return
        bucket.append([key, value])


    def get(self, key: int) -> int:
        for element in self.getBucket(key):
            if element[0] == key:
                return element[1]
        return -1


    def remove(self, key: int) -> None:
        bucket = self.getBucket(key)
        for i, e in enumerate(bucket):
            if e[0] == key:
                bucket.pop(i)
                return



# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)
